support for 'stats' unix socket

alicebob opened the pull request:
Support for HAProxy's stats socket as described in http://cbonte.github.io/haproxy-dconv/configuration-1.6.html#stats%20socket
It is enabled by using 'unix:/your/file' as the scape URI.

This saves setting up the HTTP admin website, and makes haproxy_exporter potentially easier to use with nbproc>1

alicebob commented:
For @grobie maybe?

alicebob commented:
The idea is to have the stats like this:
```
        stats socket /var/run/haproxy/stats.1.sock process 1
        stats socket /var/run/haproxy/stats.2.sock process 2
        stats socket /var/run/haproxy/stats.3.sock process 3
```

This is easier to configure than multiple statistic sites, imho.

grobie commented:
Looks great on first glance. I'll give it a thorough look tonight.

grobie commented on haproxy_exporter.go:
```diff
@@ -243,24 +247,65 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+		cmd := "show stat\n"
+		n, err := io.WriteString(f, cmd)
+		if err != nil {
+			return nil, err
```
Wouldn't it be necessary to close the connection here given an error is returned? Same below.

grobie commented:
Thanks a lot! I only have one question regarding closing connections upon write errors.

IMHO, we should have tests covering the unix connection path and an invalid URL input. Do you think you could add these?

alicebob commented:
Thanks for your quick review, @grobie!

I've added the (three) missing Close()-on-errors, and added a unittest with an actual unix socket. I guess this will break the tests on windows system (although HAProxy doesn't seem supported there it would be possible to run haproxy_exporter from a windows machine), so I could move that test to a different file which would be excluded on windows.

I can squash all the commits when done, if you want.

alicebob commented:
I've skipped the new unix socket test on windows, just to prevent some problems.

grobie commented:
Thank you for your great contribution and the tests! It looks great, please
squash.
On Jun 7, 2016 10:55, "Harmen" <notifications@github.com> wrote:

> I've skipped the new unix socket test on windows, just to prevent some
> problems.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/haproxy_exporter/pull/58#issuecomment-224220143>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAANaEe5bdUJOgyja4nl8pMyPFym_KKAks5qJTIQgaJpZM4IvFhE>
> .
>

alicebob commented:
@grobie: squashed

grobie commented:
I missed that the unix connection doesn't use a timeout. We use socket connections in an internal controller (maintenance, draining, etc.) and ran into problems there before we introduced the timeout.

I'd suggest to pass the timeout flag and use `net.DialTimeout()` and a read deadline.

alicebob commented on haproxy_exporter.go:
```diff
@@ -325,7 +329,7 @@ loop:
 		default:
 			log.Errorf("Unexpected error while reading CSV: %v", err)
-			e.csvParseFailures.Inc()
+			e.up.Set(0)
```
@grobie without this unix socket read errors would be marked as csv parse failures, which might or might not be what you want. Up to you.

grobie replied:
I think this is fine.

alicebob commented:
Added the deadline, with some tests. There is a change how read errors are counted, which I commented on inline.

grobie commented:
Awesome! Thank you so much. I'll cut a new release this week.
